awkward/array/jagged.py

```diff
@@ -47,6 +47,9 @@ def offsets2parents(cls, offsets):
 
     @classmethod
     def startsstops2parents(cls, starts, stops):
+        assert starts.shape == stops.shape
```

Comment:
Although it might never happen when calling `startsstops2parents`, it's legal for a `stops.shape[0] >= starts.shape[0]`; but `stops.shape[1:]` must be `starts.shape[1:]`.

Reply:
Is this legal because it enables a particular feature? I can't think of one myself.  Otherwise I would advocate for `_valid()` to check that condition.

Reply:
The classmethods like `startsstops2parents` are a little different because they don't own the `starts` and `stops` arrays, so it hasn't been clear to me how much they should check for validity.

The reason `stops` is allowed to be longer than `starts` (but only in the first dimension) is because it might be part of a writing process that atomically writes each array, but isn't guaranteed atomic across multiple arrays, such as `content`, `stops`, and `starts`. The policy for writing to arrays like this—which has never come up in practice—is to require writers to write inner arrays before outer arrays, so that they're always in a valid, though possibly old, state (again, assuming that arrays can be atomically extended, which only requires the "length" parameter to be an atomic integer). In the case of `JaggedArray`, which has three array children, `content` is first (innermost), followed by `stops`, followed by `starts` (outermost). The `__len__` of the `JaggedArray` is defined by `len(self._starts)` and the `stops` parameter is `self._stops[:len(self._starts)]`.

But that's not particularly relevant because awkward arrays have never been used in any application where they grow—I was just making sure that the protocol allows for growing. I think I'm going to explicitly add this when I write it in C++, because then I'll have control over the low-level arrays and can define them to be growable like the current (and currently unused) `AppendableArray`.

As for the comment on this line, I'm pointing it out because this is _stricter_ than the specification requires. These classmethod functions used to have less strict requirements.

In practical terms, I don't think it's a problem either way. This is by no means a blocking comment.